tests/utils/test_price_utils.py

```diff
@@ -0,0 +1,577 @@
+"""
+
+# pylint: disable=too-many-lines
+# pylint: disable=too-many-instance-attributes
+# pylint: disable=too-many-locals
+# pylint: disable=attribute-defined-outside-init
+
+import unittest
+import numpy as np
+
+from elfpy.utils import price as price_utils
+
+
+class TestPriceUtils(unittest.TestCase):
+    """Unit tests for price utilities"""
+
+    def test_calc_total_liquidity_from_reserves_and_price(self):
+        """
+        Test cases for calc_total_liquidity_from_reserves_and_price
+        The functions calculates total liquidity using the following formula:
+        liquidity = base_asset_reserves + token_asset_reserves * spot_price
+        """
+
+        # Test cases
+        # 1. 500k base asset, 500k pts, p = 0.95
+            # liquidity = 500000 + 500000 * 0.95 = 975000
+        # 2. 800k base asset, 200k pts, p = 0.98
+            # liquidity = 800000 + 200000 * 0.98 = 996000
+        # 3. 200k base asset, 800k pts, p = 0.92
+            # liquidity = 200000 + 800000 * 0.92 = 936000
+        # 4. 1M base asset, 0 pts, p = 1.00
+            # liquidity = 1000000 + 0 * 1.00 = 1000000
+        # 5. 0 base asset, 1M pts, p = 0.90
+            # Liquidity = 0 + 1000000 * 0.90 = 900000
+
+        test_cases = [
+            # test 1
+            {
+                "base_asset_reserves": 500000,
+                "token_asset_reserves": 500000,
+                "spot_price": 0.95,
+                "expected_result": 975000
+            },
+            # test 2
+            {
+                "base_asset_reserves": 800000,
+                "token_asset_reserves": 200000,
+                "spot_price": 0.98,
+                "expected_result": 996000
+            },
+            # test 3
+            {
+                "base_asset_reserves": 200000,
+                "token_asset_reserves": 800000,
+                "spot_price": 0.92,
+                "expected_result": 936000
+            },
+            # test 4
+            {
+                "base_asset_reserves": 1000000,
+                "token_asset_reserves": 0,
+                "spot_price": 1.00,
+                "expected_result": 1000000
+            },
+            # test 5
+            {
+                "base_asset_reserves": 0,
+                "token_asset_reserves": 1000000,
+                "spot_price": 0.90,
+                "expected_result": 900000
+            }
+        ]
+
+        for test_case in test_cases:
+            liquidity = price_utils.calc_total_liquidity_from_reserves_and_price(
+                test_case["base_asset_reserves"],
+                test_case["token_asset_reserves"],
+                test_case["spot_price"]
+            )
+            assert liquidity == test_case["expected_result"], \
+                f"Calculated liquidity doesn't match the expected amount for these inputs: {test_case}"
+
+
+    def test_calc_base_asset_reserves(self):
+        """Unit tests for the calc_base_asset_reserves function"""
+
+        test_cases = [
```

Comment:
we should always be testing unhappy paths.  can we can a test where APR is 0, 100 and > 100%?

Reply:
I think this is resolved? @sentilesdal or @MazyGio can you resolve it if you're happy?

Reply:
Added error cases, resolving this now. Stumbled upon negative results though under some edge-case-ish situations where fixed APR is very low and share_price is higher than init_share_price (i.e. users bought up all the PTs). The code runs but we might want to check the math for those cases